### PR TITLE
8329858: G1: Make G1VerifyLiveAndRemSetClosure stateless

### PR DESCRIPTION
**Pull request by stefank**
In [JDK-8329570](https://bugs.openjdk.org/browse/JDK-8329570): 'G1: Excessive is_obj_dead_cond calls in verification' we moved the verification of the containing object so that it was done only once. We left the repeated null-check because of concerns that moving it would introduce the risk that someone in the future would miss calling `set_containing_obj` or calling it with a null object.

I propose that we restructure the code to make the closure's state const and remove the set_containing_obj function so that we can feel confident that we don't hit such a bug in the future.

The patch also combines the failures from verification of the oops with failures from checking the containing objects. I had a version that kept these two separate, but I don't see why we would want that. Tell me if you'd want me to reintroduce this separation.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329858](https://bugs.openjdk.org/browse/JDK-8329858): G1: Make G1VerifyLiveAndRemSetClosure stateless (**Enhancement** - P4)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [2a009216](https://git.openjdk.org/jdk/pull/18677/files/2a009216653acc266644278bc2805c56dcbe7617)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [2a009216](https://git.openjdk.org/jdk/pull/18677/files/2a009216653acc266644278bc2805c56dcbe7617)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18677/head:pull/18677` \
`$ git checkout pull/18677`

Update a local copy of the PR: \
`$ git checkout pull/18677` \
`$ git pull https://git.openjdk.org/jdk.git pull/18677/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18677`

View PR using the GUI difftool: \
`$ git pr show -t 18677`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18677.diff">https://git.openjdk.org/jdk/pull/18677.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18677#issuecomment-2042648481)

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1HeapRegion.cpp [hunk 1]
```diff
@@ -611,42 +629,36 @@ class G1VerifyLiveAndRemSetClosure : public BasicOopIterateClosure {
...
-
-  void set_containing_obj(oop const obj) {
-    assert(!G1CollectedHeap::heap()->is_obj_dead_cond(obj, _vo), "Precondition");
-    _containing_obj = obj;
+  G1VerifyLiveAndRemSetClosure(oop containing_obj, VerifyOption vo, G1VerifyFailureCounter* failures)
+    : _vo(vo),
+      _containing_obj(containing_obj),
+      _failures(failures) {
+    assert(containing_obj != nullptr, "must be");
+    assert(!G1CollectedHeap::heap()->is_obj_dead_cond(containing_obj, _vo), "Precondition");
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1HeapRegion.cpp:651]: In this new structure, these two asserts are not very useful any more, IMO.

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1HeapRegion.cpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/gc/g1/g1HeapRegion.cpp:651]: Good point. The `is_obj_dead_cond` has already been checked by the caller. I can see why someone would like to keep the null-check. I'll see what @tschatzl says, and will make changes after that.

Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1HeapRegion.cpp:651]: Feel free to drop both.

**Review by tschatzl — APPROVED** [review]

**Comment by stefank** [comment]
Thanks for the reviews!
/integrate

**Pull request closed without merging**